### PR TITLE
 Introduce matrix for unit tests.

### PR DESCRIPTION
**Pull request by sairamsrinivasan**
Changes -

- Set up matrix in CI to run unit tests
- Add exclusion rules for incompatible versions

### CODE REVIEW

**Comment by sairamsrinivasan** [comment]
@dblock - Can you please review when you get the chance? Thanks!

**Comment by sairamsrinivasan** [comment]
> I am open to this, but why is it better? Looks like we're now maintaining a long list of excludes instead of just saying what we want to run.

My understanding is that the matrix allows us to define combinations of dependencies that may be compatible with each other that we can easily test. Manually adding a combination of gems to an even longer list can be difficult to manage. The caveat is that some gem versions can be incompatible with others, which can be easily identified by a failing build pipeline and excluded in subsequent runs.

Another alternative that I'd like to propose is that we trim down the list of supported versions that we need to use for the test. Mongoid's documentation page only supports notes for v7.3+ - which is already more than 3 years old. I have to dig up the release notes on github to find out more about older versions.

Can we test only the following versions:

MongoDB 5+
Ruby 2.6+
Mongoid 7+?

What are your thoughts about this?

**Comment by dblock** [comment]
I would prefer to leave the existing definition rather than exclusions. It's very confusing to me what is actually run with the matrix-like definition.

On the versions I only deprecate things if that's actually useful. For example, if additional work is required to maintain ruby 2.6 compatibility or code can be deleted if we cut it, go for it. But just because mongoid 6 is old doesn't mean we should be causing users pain. I interacted with a project using mongoid 6 as recently as this week.

Always open to be convinced otherwise! Please do insist if you think something is worth it. And thanks so much for your contributions and work here, I sincerely appreciate it.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -5,79 +5,64 @@ on: [push, pull_request]
...
+          - { ruby: "3.0", gemfile: "mongoid_4" }
+          - { ruby: "3.0", gemfile: "mongoid_5" }
+          - { ruby: "3.0", gemfile: "mongoid_6" }
+          - { ruby: "3.1", gemfile: "mongoid_3" }
+          - { ruby: "3.1", gemfile: "mongoid_4" }
+          - { ruby: "3.1", gemfile: "mongoid_5" }
+          - { ruby: "3.1", gemfile: "mongoid_6" }
     steps:
-      - name: Set up MongoDB ${{ matrix.entry.mongodb }}
+      - name: Cache Gems
```
Comment (dblock) [inline .github/workflows/test.yml:42]: This part is very useful, bring it in as a separate PR while we debate the pros/cons of matrix approaches?

**Comment by sairamsrinivasan** [comment]
> I would prefer to leave the existing definition rather than exclusions. It's very confusing to me what is actually run with the matrix-like definition.
> 
> On the versions I only deprecate things if that's actually useful. For example, if additional work is required to maintain ruby 2.6 compatibility or code can be deleted if we cut it, go for it. But just because mongoid 6 is old doesn't mean we should be causing users pain. I interacted with a project using mongoid 6 as recently as this week.
> 
> Always open to be convinced otherwise! Please do insist if you think something is worth it. And thanks so much for your contributions and work here, I sincerely appreciate it.

@dblock - Thank you for taking the time to review this PR and offer feedback. Let me think about this further to see how we can improve the existing definitions. I'm closing this PR for now.

I opened another PR that adds dependency caching. You can find the changes here: https://github.com/mongoid/mongoid-compatibility/pull/20

File: .github/workflows/test.yml (hunk 1 above)
Comment (sairamsrinivasan) [reply .github/workflows/test.yml:42]: Good idea. Will pull the changes into a separate PR. Thanks!

**Pull request closed without merging**